_posts/2016-10-26-jwt-artsy-journey.md

```diff
@@ -0,0 +1,46 @@
+---
+layout: post
+title: "JWT, Artsy Journey"
+date: 2016-10-26 11:03
+comments: true
+author: ashkan
+categories: [Auth, JWT, API]
+---
+
+At Artsy we currently have ~1730 client applications hitting our API and requesting authentication. On a successful authentication, we want to be able to provide basic user/application data in the token so clients can use it without needing to call our API. For that we want to use JWT.
+
+JWT (JSON Web Token) is a self-contained, secure way of transmitting data between applications and clients as JSON objects. Once decoded, JWT contains a JSON including 3 separate secions, header, payload and signature. Header section is used to define the type and details about the token itself. Payload seciton includes _claims_ and data provided by token issuer about the user/application, things like token expiration time, user id and etc. Signature section is used to verify the sender of the JWT and make sure token wasn't modified along the way.
+<!-- more -->
+
+# Where we were
+
+After a successful login, our API would generate our own access token. Our encrypted access token included some basic information about the user and the application using the token.
+For every authenticated request we would have to decrypt the token, make sure user and application are valid and application still has access to our API. We are already _stateless_ since we don't store tokens in our database.
+We can share access tokens between applications and then each application has to make a call to our API to get basic user data.
```

Comment:
In this manner, we avoided needing to store sessions in the database. However, client applications had to make an API call to get any information about the authenticated user.